### ISSUE TITLE
[REVIEW]: ddop: A python package for data-driven operationsmanagement

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3429]
**Submitting author:** @AndreasPhilippi (<a href="http://orcid.org/0000-0002-6508-9128">Andreas Philippi</a>)
**Repository:** <a href="https://github.com/opimwue/ddop" target ="_blank">https://github.com/opimwue/ddop</a>
**Version:** v0.6.5
**Editor:** @arfon
**Reviewer:** @kramea, @g4brielvs
**Archive:** <a href="https://doi.org/10.5281/zenodo.5577966" target="_blank">10.5281/zenodo.5577966</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/0de119f95840b69fcea94309c18058e4/status.svg)](https://joss.theoj.org/papers/0de119f95840b69fcea94309c18058e4)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/0de119f95840b69fcea94309c18058e4"><img src="https://joss.theoj.org/papers/0de119f95840b69fcea94309c18058e4/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/0de119f95840b69fcea94309c18058e4/status.svg)](https://joss.theoj.org/papers/0de119f95840b69fcea94309c18058e4)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/opimwue/ddop" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@kramea & @g4brielvs, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @arfon know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/opimwue/ddop">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@AndreasPhilippi) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/opimwue/ddop">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@AndreasPhilippi) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @kramea, @g4brielvs it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.25 s (360.6 files/s, 140925.6 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
JavaScript                      14           2289           2323           8445
HTML                            22           2891             66           7132
SVG                              1              0              0           2671
Jupyter Notebook                 2              0           3577            935
Python                          17            403           1303            929
CSS                              4            181             33            720
reStructuredText                23            245            212            239
TeX                              1              5              0             56
Markdown                         1             16              0             38
DOS Batch                        1             11              1             32
make                             1              4              7             10
YAML                             2              2              0              8
-------------------------------------------------------------------------------
SUM:                            89           6047           7522          21215
-------------------------------------------------------------------------------


Statistical information for the repository 'da0a833ad0e5f0f00fc28be4' was
gathered on 2021/06/29.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
AndreasPhilippi                 80       1316689        1301009          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
AndreasPhilippi           15692            1.2         10.5               15.75
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1287/opre.2018.1757 is OK
- 10.1287/mnsc.2018.3253 is OK
- 10.1080/24725854.2019.1632502 is OK
- 10.3390/info11020108 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03429/joss.03429/10.21105.joss.03429.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03429/joss.03429/10.21105.joss.03429.pdf) :page_facing_up: :point_left:

**Comment by arfon** [comment]
@kramea, @g4brielvs – This is the review thread for the paper. All of our communications will happen here from now on.

Please read the "Reviewer instructions & questions" in the first comment above.

Both reviewers have checklists at the top of this thread (in that first comment) with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention https://github.com/openjournals/joss-reviews/issues/3429 so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for the review process to be completed within about 4-6 weeks but please make a start well ahead of this as JOSS reviews are by their nature iterative and any early feedback you may be able to provide to the author will be very helpful in meeting this schedule.

**Comment by whedon** [comment]
:wave: @g4brielvs, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @kramea, please update us on how your review is going (this is an automated reminder).

**Comment by arfon** [comment]
FYI I just emailed @kramea to see when they might be able to complete their review by.

**Comment by kramea** [comment]
I will finish the review by this weekend.

**Comment by kramea** [comment]
I cannot add my review anymore. It says my invitation was revoked.

**Comment by arfon** [comment]
@whedon re-invite @kramea as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@kramea please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by arfon** [comment]
@kramea – could you try again now please? The invites from GitHub seem to expire after some time.

**Comment by kramea** [comment]
Yes, works now. Thanks!

**Comment by kramea** [comment]
Here is my review of the software:

**Installation:** 

- I created an empty environment, installed the requirements and then installed the ddop package. Importing the ddop package worked. 
- However, there were some issues when I tried to import ddop.newsvendor. It has a dependency of 'mpmath'. It needs to be included in the dependencies. 

**Functionalities & Tutorials:**

- I tried to follow the newsvendor tutorial with Yaz dataset. I believe their Tutorial page is a bit outdated (which is understandable considering the ongoing development). A lot of the usage has changed from the tutorials, including the names of the model algorithms. For instance, I could not find DecisionTreeWeighted Newsvendor. 
- Once I corrected the names, I was able to run the code to assess the functionality. It works as intended and claimed by the authors.
- Their API reference also includes descriptions of the commands.
- One suggestion would be to include more documentation on how to load custom dataset, as the authors claim this package can be used as a common framework for any kind of dataset. 

Paper:

- The authors have done a good job detailing the need for the package for operations management, as well as future work and references.

**Comment by AndreasPhilippi** [comment]
@kramea Thank you four your good feedback. 
 I have added 'mpmath' to the dependencies. I also updated the tutorial on the documentation website, so everything should be up to date now. Please note that we are continuously trying to extend the documentation.

**Comment by arfon** [comment]
:wave: @g4brielvs – how are you getting on with your review here? Are you waiting on input from the author (@AndreasPhilippi) or myself at this point?

**Comment by g4brielvs** [comment]
Here is my review (at last). Thousands apologies for my missing in action. 

### General checks

- The packages passes general checks. However, some of the dependencies were not included, e.g. `mpmath`.

### Documentation & Paper

- As indicated on the checklist, both document and paper pass the inspection. However, it would be beneficial to have more descriptions and expanded examples on the documentation, especially for those not familiar with the research area.

**Comment by AndreasPhilippi** [comment]
@arfon What is the next step after both kramea and g4brielvs have completed the review checklist?

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```
